testutil/tap.go

```diff
@@ -0,0 +1,89 @@
+package testutil
+
+import (
+	"fmt"
+	"strings"
+	"time"
+)
+
+// TapEvent represents a tap event
+type TapEvent struct {
+	Method     string
```

Comment:
are we planning to use these fields directly, and hence we need to make them public here?

Reply:
Yep. We may have to create expected tap events that we'll check against the received tap events using the `ValidateExpected` method. Plus the `tap_test.go` also need to create instances of `TapEvent` for assertion.